src/base/permutationsutil.h

```diff
@@ -0,0 +1,79 @@
+/*
+   Copyright 2016 Software Reliability Lab, ETH Zurich
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+ */
+
+#ifndef BASE_PERMUTATIONSUTIL_H_
+#define BASE_PERMUTATIONSUTIL_H_
+
+#include <limits>
+
+void ComputeAllPermutations(std::vector<int> v,
+    std::unordered_set<std::vector<int>>* permutations,
+    int n,
+    size_t permutations_beam_size) {
+  std::vector<size_t> count;
+  count.assign(v.size(), 0);
+  permutations->insert(v);
+  for (size_t i = 1; i < v.size();) {
+    if (count[i] < i) {
+      int swap = i % 2 * count[i];
+      int tmp = v[swap];
+      v[swap] = v[i];
+      v[i] = tmp;
+      permutations->insert(v);
+      if (permutations->size() > permutations_beam_size) {
+        break;
+      }
+      count[i]++;
+      i = 1;
+    } else {
+      count[i++] = 0;
+    }
+  }
+}
+
+void ComputeRandomPermutations(std::vector<int> v,
+    std::unordered_set<std::vector<int>>* permutations,
+    size_t permutations_beam_size,
+    size_t max_num_duplicates) {
```

Comment:
This name is confusing. You want to explain in a comment what it does and/or rename the parameter.